### DISCUSSION TITLE
ESP32-S3-DevKitC-1-N32R8V micropython adaption problems

### DISCUSSION THREAD

**Discussion by ferryman314** [micropython/micropython#10359] [ESP32]
Is there a suitable workaround to flash micropython on ESP32-S3-DevKitC-1-N32R8V ?

I tried the octal SPIRAM version ( https://micropython.org/resources/firmware/GENERIC_S3_SPIRAM_OCT-20221220-unstable-v1.19.1-782-g699477d12.bin) already. I could flash it and run python files via Thonny from the editor but I have problems to upload python files with thonny to the device's directory:

Error message in thonny:

![image](https://user-images.githubusercontent.com/77538649/209929557-ff25ce56-9521-4ca7-a071-454779c6b694.png)

Is there a micropython adaption in progress to make this board workable from a Windows PC?

I already tried to follow the workaround with esptool 4.4 beta under WSL but I did not managed to forward the ESP32 usb port from windows to WSL:  https://forum.micropython.org/viewtopic.php?f=18&t=11866

I am thankful for any advice (I recently tried the switch from Pico/Arduino to ESP32 and I think I started with the wrong ESP32 board for that, haha :) )

-Cheers :)

**Comment by robert-hh** [comment]
try using mpremote instead of Thonny.

**Reply by ferryman314** [reply]
Thanks for the fast reply :)

I tried mpremote and I tried to manipulate the directories on the EPS32 but I received the same error: OSError: [Errno 19] ENODEV

![image](https://user-images.githubusercontent.com/77538649/209883895-9cd1f3a9-35a0-4a2b-9348-f6ed89ea9980.png)

details:

ESP32-S3-DevKitC-1-N32R8V flashed with esptool 4.4 with GENERIC_S3_SPIRAM_OCT-20221220-unstable-v1.19.1-782-g699477d12.bin

**Comment by ferryman314** [comment]
so I solved the issue by following this procedure on linux-based machine that I set up for this purpose (i found that procedure here https://forum.micropython.org/viewtopic.php?f=8&t=11827):

#built requirements
sudo apt-get install git wget flex bison gperf python3 python3-pip python3-setuptools cmake ninja-build ccache libffi-dev libssl-dev dfu-util libusb-1.0-0

#allow current user read/write access to USB over serial port
#I fixed this part by chmod my user to the specific port 
sudo usermod -a -G dialout $USER

~/$ mkdir ~/esp
~/$ cd ~/esp

~/esp$ git clone --recursive -b v4.4-beta1  https://github.com/espressif/esp-idf.git
~/esp$ cd esp-idf
~/esp/esp-idf$ ./install.sh esp32s3
~/esp/esp-idf$ 
. $HOME/esp/esp-idf/export.sh
~/esp/esp-idf$ cd ..

~/esp$ mkdir mpy
~/esp$ cd mpy
~/esp/mpy$ git clone https://github.com/micropython/micropython --depth=1
~/esp/mpy$ cd ~/esp/mpy/micropython/ports/esp32/boards/

edit sdkconfig.spiram_sx
Change CONFIG_SPIRAM_MODE_QUAD=y -> CONFIG_SPIRAM_MODE_OCT=y

~/esp/mpy/micropython/ports/esp32/boards$ cd GENERIC_S3_SPIRAM/
edit mpconfigboard.cmake
comment out: boards/sdkconfig.usb -> # boards/sdkconfig.usb

~/esp/mpy/micropython/ports/esp32/boards/GENERIC_S3_SPIRAM$ cd ~/esp/mpy/micropython
~/esp/mpy/micropython$ make -C mpy-cross
~/esp/mpy/micropython$ cd ~/esp/mpy/micropython/ports/esp32/

edit Makefile
edit line: PORT ?= /dev/ttyUSB0 -> change to correct port /dev/ttyUSB0 for (uart to usb) oder /dev/ttyACM0 for (usb) 
edit line: BOARD ?= GENERIC -> BOARD ?= GENERIC_S3_SPIRAM
edit line: idf.py $(IDFPY_FLAGS) build -> idf.py $(IDFPY_FLAGS) menuconfig

~/esp/mpy/micropython/ports/esp32$ make submodules
~/esp/mpy/micropython/ports/esp32$ make

in menuconfig -> serial flasher config:
enable Octal Flash
Set Flash size to 16 (later to 32 but current;y not supported)
After flashing -> reset

in menuconfig -> Partition Table:
custom partition table = partitions-16MiB.csv

press S to save and Q to Quit

~/esp/mpy/micropython/ports/esp32$ make erase
~/esp/mpy/micropython/ports/esp32$ make deploy

~/esp/mpy/micropython/ports/esp32$ screen /dev/ttyUSB0 115200

XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

In any regard there is still the problem, that I can not access the full 32 MB of RAM with this procedure and I still was not able to find an easy solution to flash the N32R8V with micropython on a windows machine properly.
If there is a working procedure on Windows I would be glad to learn about that, and if there is none I would be thankful for a working implementation for boards like mine :)

-Cheers :)

**Reply by TheScilab** [reply]
Thanks for describing your solution. I started working on the same problem (working workflow for ESP32­-S3­-DevKitC-1 - ESP32-S3-WROOM-2 - 32MB Flash 8MB PSRAM in Windows).
Would you share your bin-file. Would be useful for comparison why my current workflow fails.
Thx

**Comment by i-am-tc** [comment]
@ferryman314 : Thank you for your post above. I too have a N32R8V board. Your post saved me a lot of time. 

I tried doing : git clone --recursive -b v4.4.3 https://github.com/espressif/esp-idf.git

And I was able to select 32MB for flash size in the "menuconfig". 

I checked it again by importing esp and checking flash_size() and it did show up "33554432" . Also managed to connect to a wifi network

For some reason, when I do a "import esp32", there is no "esp32.raw_temperature()" to be found.

**Reply by bixb922** [reply]
May be related to post #10443. Internal hall effect and temperature sensors seem to be implemented only for ESP32, not for later models.

**Comment by i-am-tc** [comment]
Take note also that in order to flash the firmware on say, Windows 10, one would need 3 files from the build process: micropython.bin , bootloader.bin , partition-table.bin + git clone of esptool.py repository

Assuming that these 3 files are in the same directory as esptool.py ... 

(1) Erase first : 

python esptool.py --chip esp32s3 --port COM8 erase_flash

(2) Then flash :

python esptool.py --chip esp32s3 -p COM8 -b 460800 --before=default_reset --after=hard_reset write_flash --flash_mode dout --flash_freq 80m --flash_size 32MB 0x0 bootloader.bin 0x10000 micropython.bin 0x8000 partition-table.bin

* Side note: the RGB LED on this board (ESP32-S3-N32R8V) is on pin 38 and should be accessed with the NeoPixel library, [reference](https://github.com/espressif/esp-idf/issues/8547#issuecomment-1185667739) . It is rather bright. 

![image](https://user-images.githubusercontent.com/1066699/211700790-7029ee67-3629-4d8d-8e28-64d081e6b81a.png)

**Reply by mikecando** [reply]
Is there anyway that someone can share the bin file that got created and works. I have done the same thing and I have the same errors. I have the esp32-devkitc-1-n with the 32 mb. This is chip esp32-s3-wroom-2

With a shared bin file, I believe I can do the following below. (Note I am on windows 10)

1)hold down boot button and plug in usb cable into the uart port
2) let go of the boot button
3) run the 2 commands below
esptool --chip esp32s3 --port COM10 erase_flash
esptool --chip esp32s3 --port COM10 write_flash -z 0 shared_bin_file.bin

Thank you so much

**Comment by rgrizzell** [comment]
ESP-IDF v5.0 includes the changes necessary to utilize the full 32MB of Flash. There's a considerable development and testing effort required to upgrade, though.
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/migration-guides/release-5.x/5.0/index.html

**Comment by dhalbert** [comment]
We just got CircuitPython running on the ESP32-S3-DevKitC-1-N32R8V. It's using ESP-IDF `release/v4.4`, updated as of the latest commits (which are from December 2022).

https://github.com/adafruit/circuitpython/pull/5999

Some changes needed in the board-specific `sdkconfig`, and using `dout` to flash the board. Our build process may not be that similar to MicroPython's.

So it's possible :smile:

**Reply by mattytrentini** [reply]
Thanks @dhalbert, good to know! I tried a week or so ago - and have been following you, @MicroDev1 and @anecdata's good work - but the necessary commits hadn't percolated back into the IDF releases. I've been building with the [release-v4.4](https://hub.docker.com/layers/espressif/idf/release-v4.4/images/sha256-543f8e59ca75ae1dca3f5aeec9ae902d73cf9dafdbd35cb74431056ef6681604?context=explore) docker image which doesn't yet have the changes; hopefully it'll be updated soon (I'm trying to avoid installing the IDF locally).

**Comment by HonestQiao** [comment]
Base on GENERIC_S3_SPIRAM_OCT, after add the following configuration, I can successfully compile and use.
```
# for N32R8V
CONFIG_ESPTOOLPY_OCT_FLASH=y
CONFIG_ESPTOOLPY_FLASHMODE_OPI=y

CONFIG_ESPTOOLPY_FLASHSIZE_32MB=y
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions-32MiB.csv"
```

**Reply by ehagerty** [reply]
@HonestQiao thank you so much for this work!
